virttest/qemu_devices/qcontainer.py

```diff
@@ -133,15 +133,18 @@ def get_qmp_cmds(qemu_binary, workaround_qemu_qmp_crash=False):
         self.__execute_qemu_last = None
         self.__execute_qemu_out = ""
         # Check whether we need to add machine_type
-        cmd = "%s -device \? 2>&1" % qemu_binary
+        cmd = ("echo -e 'quit' | %s -monitor stdio -nodefaults -nographic -S"
+               % qemu_binary)
```

Comment:
Well the usage of `-device \?` seemed safer to me. The difference is that the `-device` finishes early and does not require any explicit quit. Using the quit means the full qemu setup would needs to be performed and in case of some bug we'll end-up with a silently ignored failure. Anyway this is not hard NACK just a note for consideration.

Reply:
Oh I read the commit message only now. Anyway are you sure the `-device \?` can also be executed without the default machine type? Let me double- check... (I understand -help can but -device is more complex and is machine related...)

Reply:
Yes @ldoktor , I tested the upstream v5.1.0 qemu and our downstream qemu, they can execute "-device \?" without any machine type.
```
# ./qemu/aarch64-softmmu/qemu-system-aarch64 -device ?
Controller/Bridge/Hub devices:
name "i82801b11-bridge", bus PCI
...
```

```
# /usr/libexec/qemu-kvm -device ?
Controller/Bridge/Hub devices:
name "pci-bridge", bus PCI, desc "Standard PCI Bridge"
...
```

Reply:
Finally the compilation finished and yes it works. Interesting.